chain/epoch-manager/src/shard_tracker.rs

```diff
@@ -49,9 +56,29 @@ impl ShardTracker {
         match &self.tracked_shards_config {
             TrackedShardsConfig::NoShards => Ok(false),
             TrackedShardsConfig::AllShards => Ok(true),
+            TrackedShardsConfig::Shards(tracked_shards) => {
+                let shard_layout = self.epoch_manager.get_shard_layout(epoch_id)?;
+                if !shard_layout.shard_ids().contains(&shard_id) {
+                    return Err(EpochError::ShardingError(format!("Invalid shard id {shard_id}")));
+                }
+                if let Some(is_tracked) =
+                    self.descendant_of_tracked_shard_cache.lock().unwrap().get(&shard_id)
+                {
+                    return Ok(*is_tracked);
+                }
+
+                let is_tracked = check_if_descendant_of_tracked_shard(
+                    shard_id,
+                    &tracked_shards,
+                    &epoch_id,
+                    &self.epoch_manager,
+                )?;
+                self.descendant_of_tracked_shard_cache.lock().unwrap().insert(shard_id, is_tracked);
```

Comment:
Maybe move to a method? 

Micro observation - the check is called outside of the lock so it may be called multiple times in parallel though it should be rare.

Reply:
Moved to a method.
I was considering locking it fully, but went with this approach to keep the code simpler, as the case is rare, the calculation is relatively cheap, and locking the whole section would block other shards from reading the cache.